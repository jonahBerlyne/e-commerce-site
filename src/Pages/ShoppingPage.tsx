import React, { useState, useEffect } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import Categories from '../Components/Categories';
import itemData from '../Data/ItemData';
import "../Styles/Shop.css";

export default function ShoppingPage() {

 useEffect(() => {
   setItems(itemData);
   return () => {
     setItems([]);
   }
 }, []);

 const [items, setItems] = useState<any[]>([]);
 const [refresh, setRefresh] = useState<boolean>(false);
 const [selected, setSelected] = useState<any>("All");

 const changeCategory = (): void => {
  setItems(itemData);
  const selectBox = document.getElementById("selectBox") as HTMLSelectElement;
  const selectedValue: any = selectBox.options[selectBox.selectedIndex].value;
  setSelected(selectedValue);
 }

 useEffect(() => {
   const books = itemData.filter(item => item.category === "Books");
   const clothing = itemData.filter(item => item.category === "Clothing");
   const collection = itemData.filter(item => item.category === "Collection");
   const food = itemData.filter(item => item.category === "Food");
   const miscellaneous = itemData.filter(item => item.category === "Miscellaneous");
   const trinkets = itemData.filter(item => item.category === "Trinkets");
   if (selected === "Books") setItems(books);
   if (selected === "Clothing") setItems(clothing);
   if (selected === "Collection") setItems(collection);
   if (selected === "Food") setItems(food);
   if (selected === "Miscellaneous") setItems(miscellaneous);
   if (selected === "Trinkets") setItems(trinkets);
 }, [selected]);


 useEffect(() => {
   setRefresh(!refresh);
 }, [items]);

 const [search, setSearch] = useState<string>("");

 useEffect(() => {
  if (search !== "") {
    const filteredItems = (selected === "All" ? itemData : itemData.filter(item => item.category === selected)).filter(item => item.title.toLowerCase().includes(search.toLowerCase()) || item.category.toLowerCase().includes(search.toLowerCase()));
    if (filteredItems !== []) setItems(filteredItems);
  } else {
    selected === "All" ? setItems(itemData) : setItems(itemData.filter(item => item.category === selected));
  }
 }, [search, selected]);

 const navigate = useNavigate();

 return (
  <div className='shopping-page-container'>
   <div className="item-filters">
    <input type="search" placeholder='Search...' value={search} onChange={(e) => setSearch(e.target.value)} />
    <Categories changeCategory={changeCategory}/>
   </div>
   <div className="items-container">
    {items.map(item => { 
      return (
        <div key={item.id} className="item-container" onClick={() => navigate(`/${item.id}`)}>
            <img src={item.image} alt={item.title} className='item-img' />
            <div className="item-border"></div>
            <div className="item-info">
              <h4 className='item-title'>{item.title}</h4> 
              <h5 className='item-price'>${item.price}</h5>
            </div>
        </div>
      )}
    )}
   </div>
   {search !== "" && items.length === 0 &&
    <h3 className='results-msg'>No results matched.</h3>
   }
   {search !== "" && items.length > 0 &&
    <h3 className='results-msg'>Results have matched {items.length} {items.length === 1 ? "item" : "items"}.</h3>
   }
  </div>
 );
}