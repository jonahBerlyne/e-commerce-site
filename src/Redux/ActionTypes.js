export const ITEM_ADDED = "itemAdded";
export const ITEM_DECREASED = "itemDecreased";
export const ITEM_INCREASED = "itemIncreased";
export const ITEM_REMOVED = "itemRemoved";