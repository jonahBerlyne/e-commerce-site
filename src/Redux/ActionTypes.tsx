export const ITEM_ADDED: string = "itemAdded";
export const ITEM_DECREASED: string = "itemDecreased";
export const ITEM_INCREASED: string = "itemIncreased";
export const ITEM_REMOVED: string = "itemRemoved";
export const ITEM_SET: string =  "itemSet";