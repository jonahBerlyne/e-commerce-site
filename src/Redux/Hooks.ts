import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState } from './RootReducer';
import type { AppDispatch } from './Store';

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;