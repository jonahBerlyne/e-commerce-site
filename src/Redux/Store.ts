import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import cartReducer from './Slices/cartSlice';
import userReducer from './Slices/userSlice';

export const store = configureStore({
  reducer: {
    cart: cartReducer,
    user: userReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;