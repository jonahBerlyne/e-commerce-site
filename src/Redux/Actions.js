import * as actions from "./ActionTypes";

export const itemAdded = (id, title, image, price) => ({
  type: actions.ITEM_ADDED,
  payload: {
    id,
    title,
    image,
    price
  }
});

export const itemDecreased = (id, price) => ({
  type: actions.ITEM_DECREASED,
  payload: {
    id,
    price 
  }
});

export const itemIncreased = (id, price) => ({
  type: actions.ITEM_INCREASED,
  payload: {
    id,
    price 
  }
});

export const itemRemoved = id => ({
  type: actions.ITEM_REMOVED,
  payload: {
    id
  }
});