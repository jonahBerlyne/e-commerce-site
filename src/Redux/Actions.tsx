import * as actions from "./ActionTypes";

export const itemAdded = (id: number, title: string, image: string, price: number) => ({
  type: actions.ITEM_ADDED,
  payload: {
    id,
    title,
    image,
    price
  }
});

export const itemDecreased = (id: number, price: number) => ({
  type: actions.ITEM_DECREASED,
  payload: {
    id,
    price 
  }
});

export const itemIncreased = (id: number, price: number) => ({
  type: actions.ITEM_INCREASED,
  payload: {
    id,
    price 
  }
});

export const itemRemoved = (id: number) => ({
  type: actions.ITEM_REMOVED,
  payload: {
    id
  }
});

export const itemSet = (id: number, title: string, image: string, price: number, quantity: number) => ({
  type: actions.ITEM_SET,
  payload: {
    id,
    title,
    image,
    price,
    quantity
  }
});