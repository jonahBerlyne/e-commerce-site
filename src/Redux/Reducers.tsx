import * as actions from "./ActionTypes";

export default function reducer(state = [], action) {
 const isInCart = state.find(item => item.id === action.payload.id);

 switch (action.type) {
  case actions.ITEM_ADDED:

   return isInCart ? state.map(item => item.id === action.payload.id ? {...item, quantity: item.quantity + 1, price: parseFloat((action.payload.price * (item.quantity + 1)).toFixed(2))} : item) : [...state, {...action.payload, quantity: 1}];
   
  case actions.ITEM_DECREASED:
     
   return state.map(item => item.id === action.payload.id ? {...item, quantity: item.quantity - 1, price: parseFloat((action.payload.price - (action.payload.price * (1 / item.quantity))).toFixed(2))} : item);

  case actions.ITEM_INCREASED:
   
   return state.map(item => item.id === action.payload.id ? {...item, quantity: item.quantity + 1, price: parseFloat((action.payload.price + (action.payload.price * (1 / item.quantity))).toFixed(2))} : item);
     
  case actions.ITEM_REMOVED:

   return state.filter(item => item.id !== action.payload.id);

  case actions.ITEM_SET:

   return isInCart ? state.map(item => item) : [...state, action.payload];

  default:
   return state;
 }
}