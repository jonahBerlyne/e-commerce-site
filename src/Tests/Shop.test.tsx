import React from 'react';
import { render, screen, cleanup, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
import ShoppingPage from '../Pages/ShoppingPage';
import { BrowserRouter as Router } from "react-router-dom";
import { Provider } from "react-redux";
import { store } from "../Redux/Store";

jest.mock("../firebaseConfig", () => {
  return {
    apps: ["appTestId"]
  };
});

afterEach(cleanup);

describe("Shopping Page", () => {

 it("renders the shopping page", () => {
  const { container } = render(
   <Provider store={store}>
    <Router>
      <ShoppingPage />
    </Router>
   </Provider>
  );
  expect(container).toMatchSnapshot();
 });
});