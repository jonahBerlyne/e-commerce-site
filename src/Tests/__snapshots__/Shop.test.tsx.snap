// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shopping Page renders the shopping page 1`] = `
<div>
  <div
    class="shopping-page-container"
  >
    <div
      class="search-input-container"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <input
        class="search-input"
        placeholder="Search..."
        type="text"
        value=""
      />
    </div>
    <div
      class="items-container false"
      data-testid="Items"
    >
      <div
        class="item-container"
      >
        <img
          alt="Astronaut Pen"
          class="item-img"
          src="/Images/Astronaut_Pen.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Astronaut Pen
          </h4>
          <h5
            class="item-price"
          >
            $
            119.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Big Salad"
          class="item-img"
          src="/Images/Big_Salad.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Big Salad
          </h4>
          <h5
            class="item-price"
          >
            $
            7.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Box of Cuban Cigars"
          class="item-img"
          src="/Images/Cuban_Cigars.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Box of Cuban Cigars
          </h4>
          <h5
            class="item-price"
          >
            $
            599.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Cadillac Fleetwood"
          class="item-img"
          src="/Images/Cadillac_Fleetwood.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Cadillac Fleetwood
          </h4>
          <h5
            class="item-price"
          >
            $
            74999.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Coffee Table Book"
          class="item-img"
          src="/Images/Coffee_Table_Book.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Coffee Table Book
          </h4>
          <h5
            class="item-price"
          >
            $
            19.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Eggplant Calzone"
          class="item-img"
          src="/Images/Eggplant_Calzone.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Eggplant Calzone
          </h4>
          <h5
            class="item-price"
          >
            $
            9.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Executive Raincoat"
          class="item-img"
          src="/Images/Raincoat.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Executive Raincoat
          </h4>
          <h5
            class="item-price"
          >
            $
            119.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Festivus Celebration Kit"
          class="item-img"
          src="/Images/Festivus_Kit.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Festivus Celebration Kit
          </h4>
          <h5
            class="item-price"
          >
            $
            49.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Junior Mints"
          class="item-img"
          src="/Images/Junior_Mints.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Junior Mints
          </h4>
          <h5
            class="item-price"
          >
            $
            3.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Kenny Rogers Chicken"
          class="item-img"
          src="/Images/Kenny_Rogers_Chicken.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Kenny Rogers Chicken
          </h4>
          <h5
            class="item-price"
          >
            $
            9.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Loaf of Marble Rye"
          class="item-img"
          src="/Images/Marble_Rye_Loaf.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Loaf of Marble Rye
          </h4>
          <h5
            class="item-price"
          >
            $
            5.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Mackinaw Peaches"
          class="item-img"
          src="/Images/Mackinaw_Peaches.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Mackinaw Peaches
          </h4>
          <h5
            class="item-price"
          >
            $
            2.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Manya's Pony"
          class="item-img"
          src="/Images/Pony.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Manya's Pony
          </h4>
          <h5
            class="item-price"
          >
            $
            349.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Muffin Tops"
          class="item-img"
          src="/Images/Muffin_Tops.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Muffin Tops
          </h4>
          <h5
            class="item-price"
          >
            $
            6.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Non-Fat Yogurt"
          class="item-img"
          src="/Images/Non_Fat_Yogurt.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Non-Fat Yogurt
          </h4>
          <h5
            class="item-price"
          >
            $
            5.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Puffy Shirt"
          class="item-img"
          src="/Images/Puffy_Shirt.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Puffy Shirt
          </h4>
          <h5
            class="item-price"
          >
            $
            3.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Reverse Peephole"
          class="item-img"
          src="/Images/Reverse_Peephole.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Reverse Peephole
          </h4>
          <h5
            class="item-price"
          >
            $
            12.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Titleist Golf Balls"
          class="item-img"
          src="/Images/Titleist_Golf_Balls.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Titleist Golf Balls
          </h4>
          <h5
            class="item-price"
          >
            $
            19.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="TV Guide"
          class="item-img"
          src="/Images/TV_Guide.png"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            TV Guide
          </h4>
          <h5
            class="item-price"
          >
            $
            11.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
      <div
        class="item-container"
      >
        <img
          alt="Tweety Bird Pez Dispenser"
          class="item-img"
          src="/Images/Pez_Dispenser.jpeg"
        />
        <div
          class="item"
        >
          <h4
            class="item-title"
          >
            Tweety Bird Pez Dispenser
          </h4>
          <h5
            class="item-price"
          >
            $
            3.99
          </h5>
          <button
            class="btn btn-primary"
          >
            Add to Cart
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
